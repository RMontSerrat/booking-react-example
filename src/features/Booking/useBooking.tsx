import { BookingForm } from "@/components/BookingForm";
import { Modal } from "@/components/Modal";
import { useBookingManagement } from "@/hooks/useBookingManagement";
import { useModal } from "@/hooks/useModal";
import { useToast } from "@/hooks/useToast";
import { IBooking } from "@/interfaces/booking";
import WarningAmberIcon from "@mui/icons-material/WarningAmber";
import { Button } from "@mui/material";
import dayjs from "dayjs";
import { useCallback } from "react";

export function useBooking() {
  const { addToast } = useToast();
  const { deleteBooking } = useBookingManagement();
  const { openModal: openModalEdit, closeModal: closeModalEdit } = useModal();
  const { openModal: openModalDelete, closeModal: closeModalDelete } =
    useModal();

  const handleSuccessCreate = useCallback(() => {
    addToast("Booking created successfully", { type: "success" });
  }, [addToast]);

  const handleSuccessEdit = useCallback(() => {
    addToast("Booking edited successfully", { type: "success" });
    closeModalEdit();
  }, [addToast, closeModalEdit]);

  const handleEdit = useCallback(
    (editingBooking: IBooking) => {
      openModalEdit({
        body: (
          <BookingForm
            defaultValues={{
              checkIn: dayjs(editingBooking?.checkIn),
              checkOut: dayjs(editingBooking?.checkOut),
              id: editingBooking?.id,
            }}
            onSuccess={handleSuccessEdit}
          />
        ),
      });
    },
    [handleSuccessEdit, openModalEdit],
  );

  const confirmDelete = useCallback(
    (booking: IBooking) => {
      if (booking.id) {
        deleteBooking(booking.id);
        addToast("Booking deleted successfully", { type: "success" });
        closeModalDelete();
      }
    },
    [addToast, closeModalDelete, deleteBooking],
  );

  const handleDelete = useCallback(
    (booking: IBooking) => {
      openModalDelete({
        body: (
          <>
            <WarningAmberIcon color="warning" fontSize="large" />
            <Modal.Title>
              Are you sure you want to delete this reservation?
            </Modal.Title>
            <Modal.Actions>
              <Button onClick={closeModalDelete}>Cancel</Button>
              <Button
                variant="contained"
                color="primary"
                onClick={() => confirmDelete(booking)}
              >
                Confirm
              </Button>
            </Modal.Actions>
          </>
        ),
      });
    },
    [confirmDelete, closeModalDelete, openModalDelete],
  );

  return {
    handleSuccessCreate,
    handleEdit,
    handleDelete,
  };
}
