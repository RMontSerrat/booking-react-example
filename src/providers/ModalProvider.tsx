import { Modal } from "@/components/Modal";
import React, { createContext, useCallback } from "react";
import { atom, useRecoilState } from "recoil";

interface ModalOptions {
  body: React.ReactNode;
  onClose?: () => void;
}

interface ModalState {
  body: React.ReactNode | null;
  open: boolean;
  onClose: (() => void) | undefined;
}

const modalState = atom<ModalState>({
  key: "modalState",
  default: {
    body: null,
    open: false,
    onClose: undefined,
  },
});

export function ModalProvider({ children }: { children: React.ReactNode }) {
  const [modalInfo, setModalInfo] = useRecoilState(modalState);

  const openModal = useCallback(
    ({ body, onClose }: ModalOptions) => {
      setModalInfo({ body, open: true, onClose });
    },
    [setModalInfo],
  );

  const closeModal = useCallback(() => {
    if (modalInfo.onClose) {
      modalInfo.onClose();
    }
    setModalInfo({ body: null, open: false, onClose: undefined });
  }, [modalInfo, setModalInfo]);

  return (
    <ModalContext.Provider value={{ openModal, closeModal }}>
      {children}
      <Modal open={modalInfo.open} onClose={closeModal}>
        {modalInfo.body}
      </Modal>
    </ModalContext.Provider>
  );
}

export interface ModalContextType {
  openModal: (options: ModalOptions) => void;
  closeModal: () => void;
}

export const ModalContext = createContext<ModalContextType | undefined>(
  undefined,
);
