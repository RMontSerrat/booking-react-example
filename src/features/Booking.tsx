import React from 'react';
import { IBooking } from '@/interfaces/booking';
import { BookingCard } from '@/components/BookingCard';
import { BookingForm } from '@/components/BookingForm';
import { Header } from '@/components/Header';
import { useBooking } from '@/hooks/useBooking';

export function Booking() {
  const { deleteBooking, editBooking, bookings } = useBooking();

  return (
        <div>
            <Header>Booking List</Header>
            <BookingForm />
            <div>
                {bookings.map((booking: IBooking, index: number) => (
                    <BookingCard key={index}>
                        <BookingCard.CheckinLabel date={booking.checkIn} />
                        <BookingCard.CheckoutLabel date={booking.checkOut} />
                        <BookingCard.Actions>
                            <BookingCard.Edit onEdit={editBooking} />
                            <BookingCard.Delete onDelete={deleteBooking} />
                        </BookingCard.Actions>
                    </BookingCard>
                ))}
            </div>
        </div>
    );
};
