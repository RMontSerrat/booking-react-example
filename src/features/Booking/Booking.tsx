import { BookingCard } from "@/components/BookingCard";
import { BookingForm } from "@/components/BookingForm";
import { BookingList } from "@/components/BookingList";
import { Header } from "@/components/Header";
import { useBooking } from "@/hooks/useBooking";
import { useModal } from "@/hooks/useModal";
import { useToast } from "@/hooks/useToast";
import { IBooking } from "@/interfaces/booking";

export function Booking() {
  const { addToast } = useToast();
  const { deleteBooking, bookings } = useBooking();
  const { openModal, closeModal } = useModal();

  const handleSuccessCreate = () => {
    addToast("Booking created successfully", { type: "success" });
    closeModal();
  };

  const handleSuccessEdit = () => {
    addToast("Booking edited successfully", { type: "success" });
    closeModal();
  };

  const handleEdit = (editingBooking: IBooking) => {
    openModal({
      body: (
        <BookingForm
          defaultValues={{
            checkIn: editingBooking?.checkIn,
            checkOut: editingBooking?.checkOut,
            id: editingBooking?.id,
          }}
          onSuccess={handleSuccessEdit}
        />
      ),
    });
  };

  const handleDelete = (booking: IBooking) => {
    deleteBooking(booking);
    addToast("Booking deleted successfully", { type: "success" });
  };

  return (
    <div>
      <Header>Booking List</Header>
      <BookingForm onSuccess={handleSuccessCreate} />
      {!!bookings.length && (
        <BookingList>
          {bookings.map((booking: IBooking) => (
            <BookingCard key={booking.id}>
              <BookingCard.Content>
                <BookingCard.CheckinLabel date={booking.checkIn} />
                <BookingCard.CheckoutLabel date={booking.checkOut} />
              </BookingCard.Content>
              <BookingCard.Actions>
                <BookingCard.Edit onEdit={() => handleEdit(booking)} />
                <BookingCard.Delete onDelete={() => handleDelete(booking)} />
              </BookingCard.Actions>
            </BookingCard>
          ))}
        </BookingList>
      )}
    </div>
  );
}
