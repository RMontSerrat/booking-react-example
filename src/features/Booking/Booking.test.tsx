import { ModalProvider } from "@/providers/ModalProvider";
import { ToastProvider } from "@/providers/ToastProvider";
import { render } from "@testing-library/react";
import { RecoilRoot } from "recoil";
import { Booking } from "./Booking";

jest.mock("@/hooks/useBookingManagement", () => ({
  useBookingManagement: () => ({
    deleteBooking: jest.fn(),
    bookings: [
      {
        id: "1",
        checkIn: new Date("2022-01-01"),
        checkOut: new Date("2022-01-02"),
      },
      {
        id: "2",
        checkIn: new Date("2022-01-03"),
        checkOut: new Date("2022-01-04"),
      },
    ],
  }),
}));

jest.mock("@/hooks/useModal", () => ({
  useModal: () => ({
    openModal: jest.fn(),
  }),
}));

jest.mock("@/hooks/useToast", () => ({
  useToast: () => ({
    addToast: jest.fn(),
  }),
}));

describe("<Booking />", () => {
  it("should match snapshot", () => {
    const { asFragment } = render(
      <RecoilRoot>
        <ToastProvider>
          <ModalProvider>
            <Booking />
          </ModalProvider>
        </ToastProvider>
      </RecoilRoot>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
