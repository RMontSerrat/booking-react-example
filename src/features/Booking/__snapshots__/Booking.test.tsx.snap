// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Booking /> should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <h4
      class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
    >
      Booking List
    </h4>
    <form>
      <div
        class="MuiFormControl-root MuiTextField-root css-z3c6am-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r0:"
          id=":r0:-label"
        >
          Check-in
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="off"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r0:"
            inputmode="text"
            name="checkIn"
            placeholder="MM/DD/YYYY"
            type="text"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
          >
            <button
              aria-label="Choose date"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CalendarIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Check-in
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root css-z3c6am-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r2:"
          id=":r2:-label"
        >
          Check-out
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete="off"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r2:"
            inputmode="text"
            name="checkOut"
            placeholder="MM/DD/YYYY"
            type="text"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
          >
            <button
              aria-label="Choose date"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CalendarIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Check-out
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Submit
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </form>
    <div>
      <div
        class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            Check-in: 
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            Check-out: 
          </p>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-dnrpxu-MuiCardActions-root"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="EditIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="DeleteIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            Check-in: 
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            Check-out: 
          </p>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-dnrpxu-MuiCardActions-root"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="EditIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="DeleteIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
